storage/google_cloud/google_cloud_utils.py

```diff
@@ -84,9 +86,33 @@ def upload_file_to_blob(bucket_credentials_file_path, target_blob_url, f):
     :type target_blob_url: str
     :param f: File to upload, opened in binary mode.
     :type f: file-like
+    :param max_retries: Maximum number of times to retry uploading the file.
+    :type max_retries: int
+    :param blob_chunk_size: The chunk size to use for resumable uploads, in MiB
+    :type blob_chunk_size: float
     """
-    log.info(f"Uploading file to blob '{target_blob_url}'...")
-    storage_client = storage.Client.from_service_account_json(bucket_credentials_file_path)
-    blob = _blob_at_url(storage_client, target_blob_url)
-    blob.upload_from_file(f)
-    log.info(f"Uploaded file to blob")
+    try:
+        log.info(f"Uploading file to blob '{target_blob_url}'...")
+        storage_client = storage.Client.from_service_account_json(bucket_credentials_file_path)
+        blob = _blob_at_url(storage_client, target_blob_url)
+
+        # Check if blob_chunk_size is below the minimum threshold
+        if blob_chunk_size > 0.256:
+            blob.chunk_size = blob_chunk_size * 1024 * 1024
+        else:
+            blob.chunk_size = 0.256 * 1024 * 1024
+
+        blob.upload_from_file(f)
+        log.info(f"Uploaded file to blob")
+
+    except (ConnectionError, socket.timeout, Timeout) as ex:
+        log.warning("Failed to upload due to connection/timeout error")
+        if max_retries > 0:
+            log.info(f"Retrying up to{max_retries} more times with a reduced chunk_size of {int(round(blob_chunk_size/2))}MiB")
+            # lower the chunk size and start uploading from beginning because resumable_media requires so
+            f.seek(0)
+            upload_file_to_blob(bucket_credentials_file_path, target_blob_url, f,
+                                max_retries - 1, int(round(blob_chunk_size/2)))
+        else:
+            log.error(f"Failed to upload after retrying 3 times")
```

Comment:
Not necessarily - you could make 5 attempts if the caller says upload_file_to_blob(..., max_retries=5)

Reply:
updated thanks